// tf2-demo/build/es6/Parser/Packet/ClassInfo.js
import { logBase2 } from '../../Math';
export function ParseClassInfo(stream) {
    const count = stream.readUint16();
    const create = stream.readBoolean();
    const entries = [];
    if (!create) {
        const bits = logBase2(count) + 1;
        for (let i = 0; i < count; i++) {
            const entry = {
                classId: stream.readBits(bits),
                className: stream.readASCIIString(),
                dataTableName: stream.readASCIIString()
            };
            entries.push(entry);
        }
    }
    return {
        packetType: 'classInfo',
        number: count,
        create,
        entries
    };
}
export function EncodeClassInfo(packet, stream) {
    stream.writeUint16(packet.number);
    stream.writeBoolean(packet.create);
    if (!packet.create) {
        const bits = logBase2(packet.number) + 1;
        for (const entry of packet.entries) {
            stream.writeBits(entry.classId, bits);
            stream.writeASCIIString(entry.className);
            stream.writeASCIIString(entry.dataTableName);
        }
    }
}
