// tf2-demo/build/es6/Parser/Packet/GameEvent.js
import { GameEventValueType } from '../../Data/GameEvent';
function parseGameEvent(definition, stream) {
    const values = {};
    for (const entry of definition.entries) {
        const value = getGameEventValue(stream, entry);
        if (value !== null) {
            values[entry.name] = value;
        }
    }
    const name = definition.name;
    return {
        name,
        values
    };
}
function encodeGameEvent(event, definition, stream) {
    for (const entry of definition.entries) {
        const value = event.values[entry.name];
        if (typeof value === 'undefined') {
            throw new Error('empty event value');
        }
        else {
            encodeGameEventValue(value, stream, entry);
        }
    }
}
function getGameEventValue(stream, entry) {
    switch (entry.type) {
        case GameEventValueType.STRING:
            return stream.readUTF8String();
        case GameEventValueType.FLOAT:
            return stream.readFloat32();
        case GameEventValueType.LONG:
            return stream.readUint32();
        case GameEventValueType.SHORT:
            return stream.readUint16();
        case GameEventValueType.BYTE:
            return stream.readUint8();
        case GameEventValueType.BOOLEAN:
            return stream.readBoolean();
        case GameEventValueType.LOCAL:
            return null;
    }
}
function encodeGameEventValue(value, stream, entry) {
    switch (entry.type) {
        case GameEventValueType.STRING:
            if (typeof value !== 'string') {
                throw new Error(`Invalid value for game event, expected string got ${typeof value}`);
            }
            return stream.writeASCIIString(value);
        case GameEventValueType.FLOAT:
            if (typeof value !== 'number') {
                throw new Error(`Invalid value for game event, expected number got ${typeof value}`);
            }
            return stream.writeFloat32(value);
        case GameEventValueType.LONG:
            if (typeof value !== 'number') {
                throw new Error(`Invalid value for game event, expected number got ${typeof value}`);
            }
            return stream.writeUint32(value);
        case GameEventValueType.SHORT:
            if (typeof value !== 'number') {
                throw new Error(`Invalid value for game event, expected number got ${typeof value}`);
            }
            return stream.writeUint16(value);
        case GameEventValueType.BYTE:
            if (typeof value !== 'number') {
                throw new Error(`Invalid value for game event, expected number got ${typeof value}`);
            }
            return stream.writeUint8(value);
        case GameEventValueType.BOOLEAN:
            if (typeof value !== 'boolean') {
                throw new Error(`Invalid value for game event, expected boolean got ${typeof value}`);
            }
            return stream.writeBoolean(value);
    }
}
export function ParseGameEvent(stream, state) {
    const length = stream.readBits(11);
    const eventData = stream.readBitStream(length);
    const eventType = eventData.readBits(9);
    const definition = state.eventDefinitions.get(eventType);
    if (!definition) {
        throw new Error(`Unknown game event type ${eventType}`);
    }
    const event = parseGameEvent(definition, eventData);
    return {
        packetType: 'gameEvent',
        event: event
    };
}
export function EncodeGameEvent(packet, stream, state) {
    const lengthStart = stream.index;
    stream.index += 11;
    const eventId = state.eventDefinitionTypes.get(packet.event.name);
    if (typeof eventId === 'undefined') {
        throw new Error(`Unknown game event type ${packet.event.name}`);
    }
    const eventDataStart = stream.index;
    stream.writeBits(eventId, 9);
    const definition = state.eventDefinitions.get(eventId);
    if (typeof definition === 'undefined') {
        throw new Error(`Unknown game event type ${packet.event.name}`);
    }
    encodeGameEvent(packet.event, definition, stream);
    const eventDataEnd = stream.index;
    stream.index = lengthStart;
    stream.writeBits(eventDataEnd - eventDataStart, 11);
    stream.index = eventDataEnd;
}
