// tf2-demo/build/es6/PacketHandler/StringTable.js
export function handleStringTable(packet, state) {
    handleTable(packet.table, state);
}
export function handleStringTables(packet, state) {
    for (const table of packet.tables) {
        handleTable(table, state);
    }
}
export function handleStringTableUpdate(packet, state) {
    const updatedTable = state.stringTables[packet.tableId];
    handleStringTableEntries(updatedTable.name, packet.entries, state);
}
export function handleTable(table, state) {
    if (!state.getStringTable(table.name)) {
        state.stringTables.push(table);
    }
    handleStringTableEntries(table.name, table.entries, state);
}
function handleStringTableEntries(tableName, entries, state) {
    if (tableName === 'userinfo') {
        for (const entry of entries) {
            if (entry && entry.extraData) {
                calculateUserInfoFromEntry(entry.text, entry.extraData, state);
            }
        }
    }
    if (tableName === 'instancebaseline') {
        for (const instanceBaseLine of entries) {
            if (instanceBaseLine) {
                saveInstanceBaseLine(instanceBaseLine, state);
            }
        }
    }
}
function calculateUserInfoFromEntry(text, extraData, state) {
    if (extraData.bitsLeft > (32 * 8)) {
        const name = extraData.readUTF8String(32);
        const userId = extraData.readUint32();
        const steamId = extraData.readUTF8String();
        if (steamId) {
            let userState = state.userInfo.get(userId);
            if (!userState) {
                userState = {
                    name: '',
                    userId,
                    steamId: '',
                    entityId: 0
                };
            }
            userState.name = name;
            userState.steamId = steamId;
            userState.entityId = parseInt(text, 10) + 1;
            state.userInfo.set(userId, userState);
        }
    }
}
function saveInstanceBaseLine(entry, state) {
    if (entry.extraData) {
        const serverClassId = parseInt(entry.text, 10);
        state.staticBaselineCache.delete(serverClassId);
        state.staticBaseLines.set(serverClassId, entry.extraData);
    }
    else {
        throw new Error('Missing baseline');
    }
}
